DMTN-148.tex

```diff
@@ -50,9 +50,276 @@
+not have a clearly defined process to define, ingest, test, and
+deprecate them.  The lack of documentation on how these processes work
+has led to many implementations of similar paths, further complicating
+the issue.
+
+
+\section{The End State}
+
+The final state that we wish to get to has the following properties:
+
+\begin{itemize}
+\item All calibration products are either lsst.afw.image.Exposure
+  image data or an explicit calibration class in the \verb|ip_isr|
+  package.\footnote{This will require class reorganization to migrate
+    the Defects and Curve calibrations definitions.}
+\item All calibration products should be constructed from a DM Task in
+  the \verb|cp_pipe| package, and have the input data and
+  configuration parameters persisted with the calibration itself.
+  This guarantees that any product can be reconstructed in the future.
+  Exceptions will exist for external or lab created calibrations,
+  although the majority of used calibrations will likely be produced
+  directly.
+\item These provenance, config, and final product files may be saved
+  to an \verb|obs_CAMERA_data| package in git for convenience.  This
+  also allows these files to be migrated to other Butler repositories
+  in a consistent fashion.  Making these products human readable
+  allows the tracking of value changes with the evolution of the set
+  of exported products.
+\item All calibration products are accepted for use and have a date
+  range set manually, preventing test runs from being used
+  inadvertently.  Calibrations ingested from git should have sufficient
+  metadata that supply appropriate date ranges.
+\item Calibration information stored in a lsst.afw.cameraGeom object
+  may be used, but is always assumed to be overridden by supplied
+  calibration products.  No new calibration information should be
+  stored in these objects, and currently existing calibration
+  information should be deprecated.
+
+\end{itemize}
+
+\section{The Current State}
+
+\subsection{Standard Calibrations}
+
+The classic calibration products, \verb|BIAS|, \verb|DARK|,
+\verb|FLAT|, and \verb|FRINGE|, all write FITS files that are indexed
+in Butler's calibRegistry.sqlite3 repository database.  These files
+are created by \verb|pipe_drivers|/constructCalibs.py, and are
+ingested into the repository upon acceptance by
+\verb|pipe_tasks|/ingestCalibs.py.  This ingestion process sets the
+``validStart'' and ``validEnd'' values to restrict the usage of the
+calibration, as well as some identifying string which is used (along
+with calibration type, filter, and detector) to uniquely identify the
+calibration on disk.  We have been using the ``calibDate'', indicating
+when the calibration was constructed or ingested, as this string.  For
+calibrations that are generated by \verb|cp_pipe|, it may be better to
+use the Butler collection that the calibration was generated in (which
+will enforce uniqueness on simultaneous processing).  Calibrations
```

Comment:
We should agree to a naming convention for those collections that also guarantees uniqueness.  We'll still probably end up including a date in that, I'd guess.  I don't think `calibDate` was a good solution, but I think we didn't have a good solution because it's actually trickier than one would think to come up with one, and this just punts the problem to the name of the collection.

I'd be interested in whether some kind of counter is also desired, as I _imagine_ that will have to come from the data repository, and while I don't think that's a problem (unless it needs to be global _across_ data repositories, or DBB endpoints), it's a feature we'll need to add.